src/3D/Camera.cpp

```diff
@@ -9,28 +9,136 @@
 
 #include "Camera.h"
 #include "3D/LandIsland.h"
+#include "ECS/Components/Transform.h"
 #include "ECS/Registry.h"
+#include "ECS/Systems/DynamicsSystem.h"
 #include "Game.h"
 
 #include <glm/gtx/euler_angles.hpp>
+#include <glm/gtx/intersect.hpp>
+#include <glm/gtx/norm.hpp>
+#include <glm/gtx/spline.hpp>
+#include <optional>
 
 using namespace openblack;
 
-glm::mat4 Camera::getRotationMatrix() const
+Camera::Camera(glm::vec3 position, glm::vec3 rotation)
+    : _position(position)
+    , _rotation(glm::radians(rotation))
+    , _dv(0.0f, 0.0f, 0.0f)
+    , _dwv(0.0f, 0.0f, 0.0f)
+    , _dsv(0.0f, 0.0f, 0.0f)
+    , _ddv(0.0f, 0.0f, 0.0f)
+    , _duv(0.0f, 0.0f, 0.0f)
+    , _drv(0.0f, 0.0f, 0.0f)
+    , _projectionMatrix(1.0f)
+    , _velocity(0.0f, 0.0f, 0.0f)
+    , _hVelocity(0.0f, 0.0f, 0.0f)
+    , _rotVelocity(0.0f, 0.0f, 0.0f)
+    , _accelFactor(0.001f)
+    , _movementSpeed(4.0f)
+    , _maxMovementSpeed(0.005f)
+    , _maxRotationSpeed(0.005f)
+    , _mouseIsDown(false)
+    , _mouseIsMoving(false)
+    , _shiftHeld(false)
+{
+	FlyInit();
+}
+
+glm::mat4 Camera::GetRotationMatrix() const
 {
 	return glm::eulerAngleZXY(_rotation.z, _rotation.x, _rotation.y);
 }
 
 glm::mat4 Camera::GetViewMatrix() const
 {
-	return getRotationMatrix() * glm::translate(glm::mat4(1.0f), -_position);
+	return GetRotationMatrix() * glm::translate(glm::mat4(1.0f), -_position);
 }
 
 glm::mat4 Camera::GetViewProjectionMatrix() const
 {
 	return GetProjectionMatrix() * GetViewMatrix();
 }
 
+const std::optional<ecs::components::Transform> Camera::RaycastCamToLand()
+{
+	// get the hit by raycasting to the land down the camera ray
+	ecs::components::Transform intersectionTransform;
+	float intersectDistance = 0.0f;
+	glm::vec3 viewVec = GetForward();
+	auto& dynamicsSystem = Game::instance()->GetDynamicsSystem();
+	if (auto hit = dynamicsSystem.RayCastClosestHit(_position, viewVec, 1e10f))
+	{
+		intersectionTransform = hit->first;
+		return std::make_optional(intersectionTransform);
+	}
+	else if (glm::intersectRayPlane(_position, viewVec, glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(0.0f, 1.0f, 0.0f),
+	                                intersectDistance))
+	{
+		intersectionTransform.position = _position + viewVec * intersectDistance;
+		intersectionTransform.rotation = glm::mat3(1.0f);
+		return std::make_optional(intersectionTransform);
+	}
+	// else // point 100 away from camera to rotate around etc. not currently used
+	//{
+	// intersectionTransform.position = _position + viewVec * 100.f;
+	// intersectionTransform.rotation = glm::mat3(1.0f);
+	// return std::make_optional(intersectionTransform);
+	//}
```

Comment:
Are you sure you didn't mean to delete this bit of code?
Is it a To-do?

Reply:
deleted